custom_components/enphase_envoy_custom/envoy_reader.py

```diff
@@ -705,7 +705,20 @@ async def inverters_production(self):
             return None
 
         return response_dict
-
+    
+    async def daily_net_import_energy(self):
+        return (await self.daily_consumption() - await self.daily_production())
+    
+    async def daily_grid_import_energy(self):
+        return max(0, await self.daily_consumption() - await self.daily_production())
+    async def daily_grid_export_energy(self):
+        return max(0, await self.daily_production() - await self.daily_consumption())
+        
+    async def grid_import_power(self):
+        return max(0, await self.consumption() - await self.production())
+    async def grid_export_power(self):
+        return max(0, await self.production() - await self.consumption())
+    
```

Comment:
This belongs in the integration's sensors, not in the envoy reader. Envoy reader is just for reading from the API, not for any logic. Logic should be performed in a sensor class.

You also need to make a note that these entities are calculated and are *not* accurate for battery installations. I myself have a battery installation and these numbers are not useful for me.

Reply:
While you're in there, you can also try to create the calculations for your integral sensors in the integration itself. I gave this a try myself but didn't have the time to logic through it. You'll need to combine Coordinator entities with entities using `calc_change` like `BatteryEnergyChangeEntity`.